### PR TITLE
♻️ Modular oracles

### PR DESCRIPTION
**Pull request by Thomas-Smets**
Made a POC to discuss tomorrow.
In a second stage I will move the code of the OracleHub to the MainRegistry.

### CODE REVIEW

File: script/Deploy_accounts_base.sol [hunk 1]
```diff
@@ -148,32 +81,55 @@ contract ArcadiaAccountDeployment is Test {
...
-        standardERC20PricingModule.addAsset(DeployAddresses.comp_base, oracleCompToUsdArr);
-        standardERC20PricingModule.addAsset(DeployAddresses.dai_base, oracleDaiToUsdArr);
-        standardERC20PricingModule.addAsset(DeployAddresses.weth_base, oracleEthToUsdArr);
-        standardERC20PricingModule.addAsset(DeployAddresses.usdc_base, oracleUsdcToUsdArr);
-        standardERC20PricingModule.addAsset(DeployAddresses.cbeth_base, oracleCbethToEthToUsdArr);
-        standardERC20PricingModule.addAsset(DeployAddresses.reth_base, oracleRethToEthToUsdArr);
+        mainRegistry.addOracleModule(address(chainlinkOM));
+
+        oracleCompToUsdId = uint80(chainlinkOM.addOracle(DeployAddresses.oracleCompToUsd_base, "COMP", "USD"));
+        oracleDaiToUsdId = uint80(chainlinkOM.addOracle(DeployAddresses.oracleDaiToUsd_base, "COMP", "DAI"));
```
Comment (j-vp) [inline script/Deploy_accounts_base.sol:100]:
```solidity
suggestion
    oracleDaiToUsdId = uint80(chainlinkOM.addOracle(DeployAddresses.oracleDaiToUsd_base, "DAI", "USD"));
```

File: script/Deploy_accounts_base.sol [hunk 2]
```diff
@@ -148,32 +81,55 @@ contract ArcadiaAccountDeployment is Test {
...
+            DeployAddresses.dai_base, BitPackingLib.pack(BA_TO_QA_SINGLE, oracleDaiToUsdArr)
+        );
+        standardERC20PricingModule.addAsset(
+            DeployAddresses.weth_base, BitPackingLib.pack(BA_TO_QA_SINGLE, oracleEthToUsdArr)
+        );
+        standardERC20PricingModule.addAsset(
+            DeployAddresses.usdc_base, BitPackingLib.pack(BA_TO_QA_SINGLE, oracleUsdcToUsdArr)
+        );
+        standardERC20PricingModule.addAsset(
+            DeployAddresses.cbeth_base, BitPackingLib.pack(BA_TO_QA_SINGLE, oracleCbethToEthToUsdArr)
```
Comment (j-vp) [inline script/Deploy_accounts_base.sol:128]: must this be `BA_TO_QA_DOUBLE`?

File: script/Deploy_accounts_base.sol [hunk 3]
```diff
@@ -148,32 +81,55 @@ contract ArcadiaAccountDeployment is Test {
...
+        );
+        standardERC20PricingModule.addAsset(
+            DeployAddresses.reth_base, BitPackingLib.pack(BA_TO_QA_SINGLE, oracleRethToEthToUsdArr)
```
Comment (j-vp) [inline script/Deploy_accounts_base.sol:131]: must this be `BA_TO_QA_DOUBLE`?

File: script/TransferOwnership.s.sol [hunk 4]
```diff
@@ -9,32 +9,32 @@ import { ArcadiaAddresses, ArcadiaContractAddresses } from "./Constants/Transfer
...
+    StandardERC20PricingModule internal standardERC20PricingModule;
+    ChainlinkOracleModule internal chainlinkOM;
+    ILiquidator internal liquidator;
 
     constructor() {
         factory = Factory(ArcadiaContractAddresses.factory);
-        oracleHub = OracleHub(ArcadiaContractAddresses.oracleHub);
         mainRegistry = MainRegistry(ArcadiaContractAddresses.mainRegistry);
         standardERC20PricingModule = StandardERC20PricingModule(ArcadiaContractAddresses.standardERC20PricingModule);
+        chainlinkOM = ChainlinkOracleModule(ArcadiaContractAddresses.standardERC20PricingModule);
```
Comment (j-vp) [inline script/TransferOwnership.s.sol:27]:
```suggestion
        chainlinkOM = ChainlinkOracleModule(ArcadiaContractAddresses.chainlinkOM);
```

File: script/TransferOwnership.s.sol [hunk 5]
```diff
@@ -9,32 +9,32 @@ import { ArcadiaAddresses, ArcadiaContractAddresses } from "./Constants/Transfer
...
         liquidator = ILiquidator(ArcadiaContractAddresses.liquidator);
```
Comment (j-vp) [inline script/TransferOwnership.s.sol:28]: Here we use an interface, in the others the contract. Although not really related to this PR and this script will need some additional care & attention (created AF-365).

File: src/interfaces/IOracleModule.sol [hunk 6]
```diff
@@ -0,0 +1,19 @@
+/**
+ * Created by Pragma Labs
+ * SPDX-License-Identifier: MIT
+ */
+pragma solidity 0.8.19;
+
+interface IOracleModule {
+    function assetPair(uint256 oracleId) external view returns (bytes16, bytes16);
```
Comment (j-vp) [inline src/interfaces/IOracleModule.sol:8]: natspec on all functions

File: test/fuzz/libraries/BitPackingLib/Unpack.fuzz.t.sol [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    /////////////////////////////////////////////////////////////// */
+
+    function setUp() public override {
+        BitPackingLib_Fuzz_Test.setUp();
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              TESTS
+    //////////////////////////////////////////////////////////////*/
+    function testFuzz_Success_unpack(uint256 length, bool[3] memory baseToQuoteAsset, uint80[3] memory oracles)
```
Comment (j-vp) [inline test/fuzz/libraries/BitPackingLib/Unpack.fuzz.t.sol:24]: also packing test needed

Comment (Thomas-Smets) [reply test/fuzz/libraries/BitPackingLib/Unpack.fuzz.t.sol:24]: Packing is not used in any contract though, only for the tests

File: src/oracle-modules/interfaces/IMainRegistry.sol [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+ * SPDX-License-Identifier: MIT
+ */
+pragma solidity 0.8.19;
+
+interface IMainRegistry {
+    /**
+     * @notice Adds a new oracle to the Main Registry.
+     * @return oracleId Unique identifier of the oracle.
+     */
+    function addOracle() external returns (uint256 oracleId);
```
Comment (j-vp) [inline src/oracle-modules/interfaces/IMainRegistry.sol:12]:
```suggestion
    /**
     * @notice Adds a new oracle to the Main Registry.
     * @param oracleModule the address of the oracle module to add.
     * @return oracleId Unique identifier of the oracle.
     */
    function addOracle(address oracleModule) external returns (uint256 oracleId);
```

Comment (j-vp) [reply src/oracle-modules/interfaces/IMainRegistry.sol:12]: (this file is not used, it's taken from src/interface/IMainRegistry)

File: src/oracle-modules/AbstractOracleModule.sol [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+pragma solidity 0.8.19;
+
+import { Owned } from "../../lib/solmate/src/auth/Owned.sol";
+
+/**
+ * @title Abstract Oracle Module
+ * @author Pragma Labs
+ * @notice Abstract contract with the minimal implementation of an Oracle Module.
```
Comment (j-vp) [inline src/oracle-modules/AbstractOracleModule.sol:12]: Also explanation what an oracle module is and what it is meant to do?

File: src/oracle-modules/AbstractOracleModule.sol [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+    /* //////////////////////////////////////////////////////////////
+                                MODIFIERS
+    ////////////////////////////////////////////////////////////// */
+
+    /**
+     * @dev Only the Main Registry can call functions with this modifier.
+     */
+    modifier onlyMainReg() {
+        require(msg.sender == MAIN_REGISTRY, "APM: ONLY_MAIN_REGISTRY");
```
Comment (j-vp) [inline src/oracle-modules/AbstractOracleModule.sol:48]:
```suggestion
        require(msg.sender == MAIN_REGISTRY, "AOM: ONLY_MAIN_REGISTRY");
```

File: src/pricing-modules/interfaces/IMainRegistry.sol [hunk 11]
```diff
@@ -8,12 +8,6 @@ import { IPricingModule } from "../../interfaces/IPricingModule.sol";
 import { RiskModule } from "../../RiskModule.sol";
```
Comment (j-vp) [inline src/pricing-modules/interfaces/IMainRegistry.sol:8]: IPricingModule not used (and can be deleted (?) if not deleted)

File: src/pricing-modules/AbstractPricingModule.sol [hunk 12]
```diff
@@ -81,6 +81,8 @@ abstract contract PricingModule is Owned, IPricingModule {
      * @param assetId The Id of the asset.
      * @return key The unique identifier.
      * @dev Unsafe cast from uint256 to uint96, use only when the id's of the assets cannot exceed type(uint96).max.
+     * For asset where the id can be bigger as a uint96, use a mapping of asset and assetId to storage,
```
Comment (j-vp) [inline src/pricing-modules/AbstractPricingModule.sol:84]: IMainreg not used in this contract

File: src/pricing-modules/FloorERC721PricingModule.sol [hunk 13]
```diff
@@ -50,48 +49,32 @@ contract FloorERC721PricingModule is PrimaryPricingModule {
...
+     * packed in a single bytes32 object.
      */
-    function addAsset(address asset, uint256 idRangeStart, uint256 idRangeEnd, address[] calldata oracles)
+    function addAsset(address asset, uint256 idRangeStart, uint256 idRangeEnd, bytes32 oracleSequence)
         external
         onlyOwner
     {
-        // View function, reverts in OracleHub if sequence is not correct.
-        IOraclesHub(ORACLE_HUB).checkOracleSequence(oracles, asset);
+        require(idRangeStart < idRangeEnd, "PM721_AA: Invalid Range");
```
Comment (j-vp) [inline src/pricing-modules/FloorERC721PricingModule.sol:59]: could also be `<=` if for some reason we only want to allow a single ID?

File: src/pricing-modules/StandardERC20PricingModule.sol [hunk 14]
```diff
@@ -4,46 +4,27 @@
...
-    // Map asset => assetInformation.
-    mapping(address => AssetInformation) public assetToInformation;
-
-    // Struct with additional information for a specific asset.
-    struct AssetInformation {
-        uint64 assetUnit; // The unit of the asset, equal to 10^decimals.
-        address[] oracles; // Array of contract addresses of oracles.
-    }
-
+contract StandardERC20PricingModule is PrimaryPricingModule {
```
Comment (j-vp) [inline src/pricing-modules/StandardERC20PricingModule.sol:18]: no more `contract Contract is IContract`?

File: test/fuzz/MainRegistry/AddOracle.fuzz.t.sol [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+    //////////////////////////////////////////////////////////////*/
+    function testFuzz_Revert_addOracle_NonPricingModule(address unprivilegedAddress_) public {
+        vm.assume(!mainRegistryExtension.isOracleModule(unprivilegedAddress_));
+
+        vm.prank(users.creatorAddress);
+        vm.expectRevert("MR: Only OracleMod.");
+        mainRegistryExtension.addOracle();
+    }
+
+    function testFuzz_Success_addOracle(address oracleModule, uint256 oracleCounterLast) public {
```
Comment (j-vp) [inline test/fuzz/MainRegistry/AddOracle.fuzz.t.sol:32]: input oracleModule overwrites `MainRegistry_Fuzz_Test.oracleModule`

File: test/fuzz/MainRegistry/AddOracleModule.fuzz.t.sol [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    /////////////////////////////////////////////////////////////// */
+
+    function setUp() public override {
+        MainRegistry_Fuzz_Test.setUp();
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                              TESTS
+    //////////////////////////////////////////////////////////////*/
+    function testFuzz_Revert_addOracleModule_NonOwner(address unprivilegedAddress_, address oracleModule) public {
```
Comment (j-vp) [inline test/fuzz/MainRegistry/AddOracleModule.fuzz.t.sol:24]: same naming, add `_``

File: test/fuzz/MainRegistry/_MainRegistry.fuzz.t.sol [hunk 17]
```diff
@@ -9,6 +9,7 @@ import { Fuzz_Test, Constants } from "../Fuzz.t.sol";
 import { AccountV1 } from "../../../src/AccountV1.sol";
```
Comment (j-vp) [inline test/fuzz/MainRegistry/_MainRegistry.fuzz.t.sol:9]: import not used

File: test/fuzz/oracle-modules/ChainlinkOracleModule/_ChainlinkOracleModule.fuzz.t.sol [hunk 18]
```diff
@@ -0,0 +1,30 @@
+/**
+ * Created by Pragma Labs
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+pragma solidity 0.8.19;
+
+import { Fuzz_Test, Constants } from "../../Fuzz.t.sol";
```
Comment (j-vp) [inline test/fuzz/oracle-modules/ChainlinkOracleModule/_ChainlinkOracleModule.fuzz.t.sol:7]: Constants not used

File: test/fuzz/pricing-modules/AbstractPricingModule/_AbstractPricingModule.fuzz.t.sol [hunk 19]
```diff
@@ -6,7 +6,6 @@ pragma solidity 0.8.19;
 
 import { Fuzz_Test, Constants } from "../../Fuzz.t.sol";
```
Comment (j-vp) [inline test/fuzz/pricing-modules/AbstractPricingModule/_AbstractPricingModule.fuzz.t.sol:7]: Constants not used

File: test/fuzz/pricing-modules/FloorERC1155PricingModule/_FloorERC1155PricingModule.fuzz.t.sol [hunk 20]
```diff
@@ -6,7 +6,7 @@ pragma solidity 0.8.19;
 
 import { Fuzz_Test, Constants } from "../../Fuzz.t.sol";
```
Comment (j-vp) [inline test/fuzz/pricing-modules/FloorERC1155PricingModule/_FloorERC1155PricingModule.fuzz.t.sol:7]: Constants not used

File: test/fuzz/pricing-modules/FloorERC721PricingModule/_FloorERC721PricingModule.fuzz.t.sol [hunk 21]
```diff
@@ -6,7 +6,7 @@ pragma solidity 0.8.19;
 
 import { Fuzz_Test, Constants } from "../../Fuzz.t.sol";
```
Comment (j-vp) [inline test/fuzz/pricing-modules/FloorERC721PricingModule/_FloorERC721PricingModule.fuzz.t.sol:7]: Constants not used

File: test/fuzz/pricing-modules/StandardERC20PricingModule/_StandardERC20PricingModule.fuzz.t.sol [hunk 22]
```diff
@@ -6,8 +6,7 @@ pragma solidity 0.8.19;
 
 import { Fuzz_Test, Constants } from "../../Fuzz.t.sol";
```
Comment (j-vp) [inline test/fuzz/pricing-modules/StandardERC20PricingModule/_StandardERC20PricingModule.fuzz.t.sol:7]: Constants not used

File: test/fuzz/pricing-modules/StandardERC4626PricingModule/_StandardERC4626PricingModule.fuzz.t.sol [hunk 23]
```diff
@@ -7,7 +7,6 @@ pragma solidity 0.8.19;
 import { Fuzz_Test, Constants } from "../../Fuzz.t.sol";
```
Comment (j-vp) [inline test/fuzz/pricing-modules/StandardERC4626PricingModule/_StandardERC4626PricingModule.fuzz.t.sol:7]: Constants not used

File: test/fuzz/pricing-modules/UniswapV2PricingModule/_UniswapV2PricingModule.fuzz.t.sol [hunk 24]
```diff
@@ -10,8 +10,8 @@ import { ERC20 } from "../../../../lib/solmate/src/tokens/ERC20.sol";
 import { StdStorage, stdStorage } from "../../../../lib/forge-std/src/Test.sol";
```
Comment (j-vp) [inline test/fuzz/pricing-modules/UniswapV2PricingModule/_UniswapV2PricingModule.fuzz.t.sol:10]: ERC20 not used

File: test/fuzz/pricing-modules/UniswapV3PricingModule/_UniswapV3PricingModule.fuzz.t.sol [hunk 25]
```diff
@@ -12,6 +12,7 @@ import { FixedPointMathLib } from "../../../../lib/solmate/src/utils/FixedPointM
 import { StdStorage, stdStorage } from "../../../../lib/forge-std/src/Test.sol";
 
 import { ArcadiaOracle } from "../../../utils/mocks/ArcadiaOracle.sol";
+import { BitPackingLib } from "../../../../src/libraries/BitPackingLib.sol";
 import { ERC20Mock } from "../../../utils/mocks/ERC20Mock.sol";
```
Comment (j-vp) [inline test/fuzz/pricing-modules/UniswapV3PricingModule/_UniswapV3PricingModule.fuzz.t.sol:16]: ERC20Mock not used

File: test/fuzz/Fuzz.t.sol [hunk 26]
```diff
@@ -5,9 +5,9 @@
 pragma solidity 0.8.19;
 
 import { Base_Test, Constants } from "../Base.t.sol";
+import { BitPackingLib } from "../../src/libraries/BitPackingLib.sol";
 import { MockOracles, MockERC20, MockERC721, MockERC1155, Rates } from "../utils/Types.sol";
 import { MainRegistry } from "../../src/MainRegistry.sol";
-import { OracleHub } from "../../src/OracleHub.sol";
 import { PricingModule } from "../../src/pricing-modules/AbstractPricingModule.sol";
 import { TrustedCreditorMock } from "../utils/mocks/TrustedCreditorMock.sol";
 import { Proxy } from "../../src/Proxy.sol";
```
Comment (j-vp) [inline test/fuzz/Fuzz.t.sol:13]: Proxy not used

File: test/Base.t.sol [hunk 27]
```diff
@@ -9,9 +9,9 @@ import { Users, MockOracles, MockERC20, MockERC721, Rates } from "./utils/Types.
 import { Factory } from "../src/Factory.sol";
```
Comment (j-vp) [inline test/Base.t.sol:9]: MockOracles, MockERC20, MockERC721, Rates, ERC20Mock, ERC721Mock, ERC1155Mock not used

File: script/TransferOwnership.s.sol (hunk 5 above)
Comment (Thomas-Smets) [reply script/TransferOwnership.s.sol:28]: Liquidator should be deployed in lending?

File: src/pricing-modules/StandardERC20PricingModule.sol (hunk 14 above)
Comment (Thomas-Smets) [reply src/pricing-modules/StandardERC20PricingModule.sol:18]: There is no IStandardERC20PricingModule interface anymore.

**Pull request merged**